### PR TITLE
[RFC] Hide Settings implementation

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Hide Settings implementation

The idea is that you can include `Settings.h` to work with a Settings object, without the need to know about its internals, macros or settings declaration. Just declare `extern const Settings{Type} name` (as we do for ProfileEvents, ErrorCodes, etc) and use the subscript (`[]`) operator.

Major simplifications would happen in BaseSettings once we remove direct usages.

Good points:
* Setting struct can be much smaller (future)
* Way less dependencies on inner details of settings, which means less build time and less rebuilds to add a setting.

Bad points:
* Need to specify the extern of each setting used in the file. If not done compilation fail, if done incorrectly it will work and it will link but have undefined errors in runtime. This is detected by style check so it should not appear.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 19c5315548769398278fa91faf211ce1a5549faf with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69213/19c5315548769398278fa91faf211ce1a5549faf/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/253d099f1cbc361191fdc7e0f044501751100de6/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Core/Settings.cpp [hunk 1]
```diff
@@ -165,4 +167,44 @@ void Settings::applyCompatibilitySetting(const String & compatibility_value)
...
+    const auto & accessor = SettingsTraits::Accessor::instance();
+    return accessor.getName(id);
+}
+
+template <settings_supported_types T>
+T get(const Settings & settings, size_t id)
+{
+    const auto & accessor = SettingsTraits::Accessor::instance();
+    auto field = accessor.getValue(settings, id);
+    /// TODO: Add assert for debug mode
```
Comment (antaljanosbenjamin) [inline src/Core/Settings.cpp:189]: `safeGet` has an assert for type safety, isn't it? Or what kind of assert would like to add?

**Comment by antaljanosbenjamin** [comment]
> checks happen at runtime, not compilation time.

I think this is the biggest drawback of the new approach, the rest (specifiy type, forward declare the used settings) is minor changes in my opinion. However I am happy to try this new approach, because the current approach makes horrible to edit any settings.

File: src/Core/Settings.cpp (hunk 1 above)
Comment (Algunenano) [reply src/Core/Settings.cpp:189]: I didn't check safeGet. I just leave notes to myself to review it later once the API is more clear

**Comment by nickitat** [comment]
i wonder how much do you think could help if we just remove all `#include <Core/Settings.h>` from headers? we will still have 500+ targets that have to be rebuild on any settings change, but at least there will be no "recursive" rebuilds.

**Comment by Algunenano** [comment]
> i wonder how much do you think could help if we just remove all #include <Core/Settings.h> from headers? we will still have 500+ targets that have to be rebuild on any settings change, but at least there will be no "recursive" rebuilds.

This was mostly done in https://github.com/ClickHouse/ClickHouse/pull/66404. I've removed a few more in this PR, and will continue in other PRs with smaller changes that won't require editing ~500 files anymore.

**Comment by Algunenano** [comment]
> Integration tests (tsan) [6/6] — fail: 41, passed: 110
Flaky kafka tests. Works fine in other runs and locally

**Pull request merged**